[Lateral Movement Detection] Add dashboard and release to production

sodhikirti07 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?
This PR adds the following enhancements in the Lateral Movement Detection package:
- Dashboard
- Updated `README` with instructions to add dashboard
- Updated `changelog` and `manifest` yml files: added a new `condition` property for premium subscription, changed package `version` for production release and added `license`.
- Added `ML` tags and bumped version in security rules
<!-- Mandatory
Explain here the changes you made on the PR.
-->

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- https://github.com/elastic/security-ml/issues/203

## Screenshots
Refer to below comment for screenshots!

sodhikirti07 commented:
Lateral Movement Detection package updated for version `0.0.2` with dashboard asset:

![image](https://user-images.githubusercontent.com/109447885/211930675-61bb41f6-1324-4d1d-937c-3ed894c6c82a.png)

Anomaly detection jobs created and started:

![image](https://user-images.githubusercontent.com/109447885/211930787-4838ad00-ed76-4247-a62f-5c2b8fbe0f40.png)

Detection rules enabled and started without warnings:

![image](https://user-images.githubusercontent.com/109447885/211930879-9d889349-b74a-410c-a4b8-934192dcec85.png)

`ML` tag added to the detection rules using machine learning jobs:

![image](https://user-images.githubusercontent.com/109447885/211930980-b9bd5346-db43-46e6-a61d-dca221aab798.png)

Dashboard asset installed and available without any errors:

![image](https://user-images.githubusercontent.com/109447885/211931087-3b9b3aeb-0464-4fa6-a4b9-bdbec08ce4a9.png)

![image](https://user-images.githubusercontent.com/109447885/211931154-08fcd8c2-725a-4c26-a8ca-26d94cf7ed6d.png)

sodhikirti07 commented on packages/lmd/kibana/dashboard/lmd-17fea180-8c4c-11ed-bb03-41a73f349362.json:
```diff
@@ -0,0 +1,52 @@
+            "syncTooltips": false,
+            "useMargins": true
+        },
+        "panelsJSON": "[{\"version\":\"8.5.1\",\"type\":\"lens\",\"gridData\":{\"x\":0,\"y\":0,\"w\":18,\"h\":10,\"i\":\"ddb33c4a-9eff-4d40-92c9-297d76d91eea\"},\"panelIndex\":\"ddb33c4a-9eff-4d40-92c9-297d76d91eea\",\"embeddableConfig\":{\"attributes\":{\"title\":\"\",\"visualizationType\":\"lnsLegacyMetric\",\"type\":\"lens\",\"references\":[{\"type\":\"index-pattern\",\"id\":\".ml-anomalies*\" ...
```
Tried to format this part, but this is how the dashboard gets exported using `elastic-package`.

szeitlin reviewed (approved):
Looks great! Ship it!

Great work getting this done so fast! 🎉

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4974/detail/PR-4974/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4974/detail/PR-4974/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4974/detail/PR-4974/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4974/detail/PR-4974/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_4974.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-01-11T22:09:42.093Z&rangeTo=2023-01-11T22:29:42.093Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=b1423c38fb860f4fe829b818e30fab2f&transactionId=e0c2abc176dc30fe)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-01-11T22:19:42.093+0000


* Duration: 12 min 30 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4974/detail/PR-4974/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: lmd`
<ul>
<li>Took 0 min 8 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-4974/runs/1/steps/1051/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-4974/runs/1/steps/1113/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-4974/runs/1/steps/1126/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

susan-shu-c reviewed (approved):
🚀

sodhikirti07 commented:
Screenshot of dashboard with data:

andrewkroh commented on packages/lmd/manifest.yml:
```diff
@@ -1,15 +1,15 @@
 categories:
   - security
-release: experimental
+release: ga
```
If you are marking this as GA then the version should move to 1.0.0 to be aligned with semantic versioning.

The `release` field is deprecated in the [package-spec](https://github.com/elastic/package-spec/blob/60560379ad6ed5ffa6fc3beb13ec22ad89bdddc4/spec/integration/manifest.spec.yml#LL346C33-L346C83) such that the version number can solely be used in determining a package's status based on semver. So I would drop the `release` field entirely and use the semver to indicate stable/GA vs dev/pre-release/beta.

sodhikirti07 replied:
Thanks for pointing. I made the changes to the file.

andrewkroh commented on packages/lmd/manifest.yml:
```diff
@@ -1,15 +1,15 @@
 format_version: 1.0.0
```
Recommend that could move to the latest `format_version: 2.2.0`. This will require that you specify the [license](https://github.com/elastic/package-spec/blob/60560379ad6ed5ffa6fc3beb13ec22ad89bdddc4/spec/integration/manifest.spec.yml#L152-L153).

sodhikirti07 replied:
Done!

peteharverson commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,13 @@ Detects potential lateral movement activity by identifying malicious file transf
+Before going to the dashboard, make sure you have the following settings configured in your Kibana. 
+1. You have started the above anomaly detection jobs.
+2. After enabling the jobs, go to **Management > Stack Management > Kibana > Data Views**. 
+3. Click on **Create data view** button and create a data view with an **Index pattern** of **.ml-anomalies***. Make sure you assign the same value to **Name** and **Custom data view ID** under **Show advanced settings**.
```
As discussed with @sodhikirti07, note that anyone viewing the dashboard will also need read access to the `.ml-anomalies*` index to be able to see the visualizations which query that index directly. To create the data view you also need to enable `Allow hidden and system indices` inside the Advanced settings.

sodhikirti07 replied:
Thank you for catching this. I've made the changes.

ajosh0504 commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,14 @@ Detects potential lateral movement activity by identifying malicious file transf
+
+The **Lateral Movement Detection Dashboard** is available under **Analytics > Dashboard**. This dashboard gives an overview of anomalies triggered for the lateral movement detection package.
+
+Before going to the dashboard, make sure you have the following settings configured in your Kibana. 
```
```suggestion
For the dashboard to work as expected, the following settings need to be configured in Kibana. 
```

sodhikirti07 replied:
Done

ajosh0504 commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,14 @@ Detects potential lateral movement activity by identifying malicious file transf
+1. You have started the above anomaly detection jobs.
+2. You have **read** access to **.ml-anomalies*** index or are assigned **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html).
+3. After enabling the jobs, go to **Management > Stack Management > Kibana > Data Views**. 
+4. Click on **Create data view** button and create a data view with an **Index pattern** of **.ml-anomalies***. Make sure you assign the same value to **Name** and **Custom data view ID** under **Show advanced settings**. You also need to enable **Allow hidden and system indices** inside the **Advanced settings**.
```
Allow hidden and system indices should happen before they can proceed to naming the pattern, so maybe that should be the first sentence in the bullet.

Also maybe explicitly mention what the Name and Data View ID must be, since I'm guessing the dashboard is relying on a specific ID.

sodhikirti07 replied:
Done

ajosh0504 commented on packages/lmd/kibana/dashboard/lmd-17fea180-8c4c-11ed-bb03-41a73f349362.json:
```diff
@@ -0,0 +1,52 @@
+    },
+    "references":  [
+    {
+      "id": ".ml-anomalies*",
```
Consider making this index pattern more specific. I believe the job results go to `.ml-anomalies-shared` by default, unless you're specifying custom indices.

ajosh0504 commented on packages/lmd/kibana/security_rule/2577e749-ed63-4b58-ac32-4f8a94d73e60.json:
```diff
@@ -16,7 +16,8 @@
             "Elastic",
             "Lateral Movement Detection",
-            "host"
+            "host",
```
Let's keep all tags in title case for consistency.
```suggestion
            "Host",
```

ajosh0504 commented on packages/lmd/kibana/security_rule/282cd7e6-9d51-421e-93b4-ebc2d422eb44.json:
```diff
@@ -16,7 +16,8 @@
             "Elastic",
             "Lateral Movement Detection",
-            "host"
+            "host",
```
```suggestion
            "Host",
```

ajosh0504 commented on packages/lmd/kibana/security_rule/2577e749-ed63-4b58-ac32-4f8a94d73e60.json:
```diff
@@ -16,7 +16,8 @@
+            "ML"
         ],
         "type": "machine_learning",
         "version": 1
```
Bump the version number since you're updating the rule.

sodhikirti07 replied:
Done

ajosh0504 commented on packages/lmd/kibana/security_rule/e5ebb55f-97f0-4815-b05b-59b223b1d4df.json:
```diff
@@ -16,7 +16,8 @@
             "Elastic",
             "Lateral Movement Detection",
-            "host"
+            "host",
```
```suggestion
            "Host",
```

sodhikirti07 commented:
@andrewkroh Thanks for mentioning this. Added the `platinum` subscription condition in `manifest.yml` and removed it from package description. @ajosh0504 @susan-shu-c We should do it for other packages as well.

![image](https://user-images.githubusercontent.com/109447885/213485294-138f7b16-84bc-4048-bff8-5916e050537d.png)

sodhikirti07 commented on packages/lmd/manifest.yml:
```diff
@@ -2,7 +2,8 @@ format_version: 2.2.0
-license: "Elastic-2.0"
+source:
+  license: "Elastic-2.0"
 description: "ML package to detect lateral movement based on file transfer activity. Requires a Platinum subscription."
```
Done!

andrewkroh reviewed (approved):
manifest.yml changes LGTM,

szeitlin commented on packages/lmd/docs/README.md:
```diff
@@ -44,9 +44,11 @@ Detects potential lateral movement activity by identifying malicious file transf
 1. You have started the above anomaly detection jobs.
-2. After enabling the jobs, go to **Management > Stack Management > Kibana > Data Views**. 
-3. Click on **Create data view** button and create a data view with an **Index pattern** of **.ml-anomalies***. Make sure you assign the same value to **Name** and **Custom data view ID** under **Show advanced settings**.
+2. You have **read** access to **.ml-anomalies*** index or are assigned **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html).
```
```suggestion
2. You have **read** access to **.ml-anomalies*** index or are assigned the **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html).
```

szeitlin commented on packages/lmd/manifest.yml:
```diff
@@ -1,15 +1,17 @@
+version: 1.0.0
+source:
+  license: "Elastic-2.0"
+description: "ML package to detect lateral movement based on file transfer activity."
```
I see `Requires a Platinum subscription` got removed <-- is this no longer true?

sodhikirti07 replied:
It is true for our package but I added it under `condition` property in `manifest.yml`. So, it will appear on right hand side of the package's main page (see the screenshot above).

szeitlin replied:
Thanks, I dislike how github doesn't have a way to insert screenshots into the code review view, so I was just looking at the code when I wrote that.

szeitlin replied:
FWIW, in the future this is the kind of thing you could put in the description of the PR (it's ok to update that as you go along).

szeitlin reviewed (approved):
lgtm!

ajosh0504 commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,15 @@ Detects potential lateral movement activity by identifying malicious file transf
+2. You have **read** access to **.ml-anomalies*** index or are assigned the **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html).
+3. After enabling the jobs, go to **Management > Stack Management > Kibana > Data Views**. 
+4. Click on **Create data view** button and enable **Allow hidden and system indices** under the **Show Advanced settings**.
+5. Create a data view with an **Index pattern** of **.ml-anomalies-shared** and assign the same index pattern to **Name** and **Custom data view ID** under **Show advanced settings**.
```
```suggestion
5. Create a data view with the following settings:
    - Index pattern : `.ml-anomalies-shared`
    - Name: `.ml-anomalies-shared`
    - Custom data view ID: `.ml-anomalies-shared`
```

droberts195 commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,15 @@ Detects potential lateral movement activity by identifying malicious file transf
+
+For the dashboard to work as expected, the following settings need to be configured in Kibana. 
+1. You have started the above anomaly detection jobs.
+2. You have **read** access to **.ml-anomalies*** index or are assigned the **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html).
```
Directly granting access to machine learning results indices has implications for using spaces to control ML job visibility. I think this should be made clear. Something along the lines of this:

```suggestion
2. You have **read** access to **.ml-anomalies*** index or are assigned the **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html). Please be aware that a user who has access to the underlying machine learning results indices can see the results of _all_ jobs in _all_ spaces. Do not grant these permissions if you use Kibana spaces to control which users can see which machine learning results. For more information on machine learning privileges see [here](https://www.elastic.co/guide/en/machine-learning/current/setup.html#setup-privileges).
```

ajosh0504 reviewed (approved):
Looking great! Nice work! 🚢

sodhikirti07 commented on packages/lmd/docs/README.md:
```diff
@@ -40,5 +40,18 @@ Detects potential lateral movement activity by identifying malicious file transf
+
+For the dashboard to work as expected, the following settings need to be configured in Kibana. 
+1. You have started the above anomaly detection jobs.
+2. You have **read** access to **.ml-anomalies*** index or are assigned the **machine_learning_user** role. For more information on roles, please refer to [Built-in roles in Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html). Please be aware that a user who has access to the underlying machine learning result's indices can see the results of _all_ jobs i ...
```
@droberts195 Thank for for your suggestion. Please review and lmk if you're okay with the text I added.

droberts195 reviewed (approved):
The ML permissions documentation LGTM.

elasticmachine commented:
Package lmd - 1.0.0 containing this change is available at https://epr.elastic.co/search?package=lmd
